lib/astc_encode.cpp

```diff
@@ -0,0 +1,963 @@
+/* vi: set sw=2 ts=4 expandtab: */
+
+/*
+ * Copyright (c) 2021, Arm Limited and Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @internal
+ * @file astc_encode.cpp
+ * @~English
+ *
+ * @brief Functions for compressing a texture to astc format.
+ *
+ * @author Wasim Abbas , www.arm.com
+ */
+
+#include <cstring>
+#include <inttypes.h>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <zstd.h>
+#include <KHR/khr_df.h>
+
+#include "dfd.h"
+#include "ktx.h"
+#include "ktxint.h"
+#include "texture2.h"
+#include "vkformat_enum.h"
+#include "vk_format.h"
+
+#include "astcenc.h"
+#include "../tools/toktx/image.hpp"
+
+// Provide pthreads support on windows
+#if defined(_WIN32) && !defined(__CYGWIN__)
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+typedef HANDLE pthread_t;
+typedef int pthread_attr_t;
+
+/* Public function, see header file for detailed documentation */
+static int
+pthread_create(pthread_t* thread, const pthread_attr_t* attribs,
+               void* (*threadfunc)(void*), void* thread_arg) {
+    (void)attribs;
+    LPTHREAD_START_ROUTINE func = (LPTHREAD_START_ROUTINE)threadfunc;
+    *thread = CreateThread(nullptr, 0, func, thread_arg, 0, nullptr);
+    return 0;
+}
+
+/* Public function, see header file for detailed documentation */
+static int
+pthread_join(pthread_t thread, void** value) {
+    (void)value;
+    WaitForSingleObject(thread, INFINITE);
+    return 0;
+}
+
+#endif
+
+static astcenc_image*
+imageAllocate(uint32_t bitness,
+              uint32_t dim_x, uint32_t dim_y, uint32_t dim_z) {
+    astcenc_image *img = new astcenc_image;
+    assert(img);
+
+    img->dim_x         = dim_x;
+    img->dim_y         = dim_y;
+    img->dim_z         = dim_z;
+
+    if (bitness == 8) {
+        void **data    = new void *[dim_z];
+        img->data_type = ASTCENC_TYPE_U8;
+        img->data      = data;
+
+        for (uint32_t z = 0; z < dim_z; z++) {
+            data[z] = new uint8_t[dim_x * dim_y * 4];
+        }
+    }
+    else if (bitness == 16) {
+        void **data    = new void *[dim_z];
+        img->data_type = ASTCENC_TYPE_F16;
+        img->data      = data;
+
+        for (uint32_t z = 0; z < dim_z; z++) {
+            data[z] = new uint16_t[dim_x * dim_y * 4];
+        }
+    }
+    else {       // if (bitness == 32)
+        assert(bitness == 32);
+        void **data    = new void *[dim_z];
+        img->data_type = ASTCENC_TYPE_F32;
+        img->data      = data;
+
+        for (uint32_t z = 0; z < dim_z; z++) {
+            data[z] = new float[dim_x * dim_y * 4];
+        }
+    }
+
+    return img;
+}
+
+static void
+imageFree(astcenc_image *img) {
+    if (img == nullptr) {
+        return;
+    }
+
+    for (uint32_t z = 0; z < img->dim_z; z++) {
+        delete[](char *) img->data[z];
+    }
+
+    delete[] img->data;
+    delete img;
+}
+
+static astcenc_image*
+unorm8x1ArrayToImage(const uint8_t *data, uint32_t dim_x, uint32_t dim_y) {
```

Comment:
I thought you were going to reuse the swizzle functions from basis_encode.cpp.

Reply:
Okay. It's not a huge deal.

> Third there are 2 functions missing in basis encode L and LA

We support these via the input swizzle that is applied before encoding. We make sure L inputs have the value in R, G & B channels and LA inputs have the L value also in all 3 channels and A in the alpha channel.